Automattic/simplenote-android
# Refactor Tag Management in NoteEditorFragment

Fixes #1478.

### Fix

This PR refactors tag management in the `NoteEditorFragment`. Since `NoteEditorFragment` is one of the more complex classes in the project, we will refactor it at different stages to use the new architecture added in #1459. In this project, we are going to extract the tag management logic into a `ViewModel` to leverage `Repositories` and `Use Cases`. 

The refactor also adds new behavior to support the collaboration UI described in #1480 (see Details section). The concrete change is that we are not going to show emails (collaborators) as tags. They will be shown in their own UI. This behavior should be implemented in all screens that show a list of tags. 

Thus, this PR adds the use cases `GetTagsUseCase` and `ValidateTagUseCase` which will be used as the single source of truth to get tags (and filter collaborators) and validate a tag. Both use cases will be used in several places in the app.

Another important change in this PR is that the method `saveNote()` in `NoteEditorFragment` does not update tags. It used the list of tags on the UI (`Chip` views) to save the tags, but since that list does not include collaborators (emails), it was deleting them when the note was closed. The tags are updated correctly in all the other actions we included. 

| Before      | After |
| ----------- | ----------- |
| ![](https://d.pr/i/TXg5j8+.gif)      | ![](https://d.pr/i/auR9c5+.gif)       |

### Test

**Note:** important to test in tablets also since the parent activity changes when in landscape mode in tablets. Use <= 2.21 to test whether the emails are added as tags.

1. Go to a note
2. Scroll down to the tags field
3. Add a tag and tap on `Done` button in the keyboard. For example `tag1`
4. Add an email. A `Toast` showing `<email> added as a collaborator.` should be shown
5. Delete a tag
6. Write to add a tag but do not tap on the button `Done`. For example `tag2`
7. Tap on the note editor. The tag should be added with whatever the tag field had
8.  Write to add a tag but do not tap on the button `Done`. For example `tag3`
9. Tab on the back button on top-left corner
10. Go to the same note again
11. The tag `tag3` should be in the list of tags

### Review

Only one developer and one designer are required to review these changes, but anyone can perform the review.


### Release

These changes do not require release notes.
